main/java/mjc/beans/AbstractPurchase.java

```diff
@@ -0,0 +1,76 @@
+package main.java.mjc.beans;
+
+import main.java.mjc.exceptions.CsvLineException;
+
+public abstract class AbstractPurchase implements Comparable<AbstractPurchase>, Cloneable {
+    private final Product product;
+    private final int amount;
+
+    public AbstractPurchase(String[] fields) throws Exception {
+        this(getValidPurchase(fields));
+    }
+
+    public AbstractPurchase(AbstractPurchase purchase) {
+        this(purchase.product, purchase.amount);
+    }
+
+    public AbstractPurchase(Product product, int amount) {
+        this.product = product;
+        this.amount = amount;
+    }
+
+    public AbstractPurchase(String name, int price, int amount) {
+        this.product = new Product(name, price);
+        this.amount = amount;
```

Comment:
bad, you give one more primary ctor. 
```suggestion
        this(new Product(name, price), amount);
```